Issue 6664: Caching of Request Ids for Request Tracker causes OOM on Controller

RaulGracia opened the pull request:
**Change log description**  
Limit max number of parallel request ids for a given request descriptor in `RequestTracker`.

**Purpose of the change**  
Fixes #6664.

**What the code does**  
Sets a limit of max parallel request ids (`MAX_PARALLEL_REQUESTS`) for a given request descriptor in `RequestTracker`. Note that the most recent `MAX_PARALLEL_REQUESTS` request ids are kept in cache.

**How to verify it**  
New tests added to verify that the limit holds. System tests are green with this PR (TPC 4432).

andreipaduroiu commented on common/src/main/java/io/pravega/common/tracing/RequestTracker.java:
```diff
@@ -170,6 +174,9 @@ public void trackRequest(String requestDescriptor, long requestId) {
+            if (requestIds.size() > MAX_PARALLEL_REQUESTS) {
+                requestIds.remove(1); // Delete the oldest parallel request id to bound the size of this list.
+            }
             ongoingRequests.put(requestDescriptor, requestIds);
```
You're putting this in the cache every time you invoke this method. This is expensive and unnecessary. Consider putting it just after line 173 (first if block).

Also consider using `SimpleCache`.

RaulGracia replied:
We need to put every new `requestId` in the cache and associate it with the `requestDescriptor`, as the intent is to track every `requestId` that attempts to do some action (in the Controller, so far). For this reason we have 1 `requestDescriptor` that can be associated to multiple concurrent `requestIds`. The only thing I have added is a limit to the amount of `requestIds` that could be associated to a particular `requestDescriptor`, given that in situations in which the Controller cannot complete a given action (e.g., Segment Store is unavailable), the `requestIds` keep accumulating for each retry, thus consuming more heap. I'm not sure if I understand your comment here (specially considering that these are constant time iserts on a map).

I have switched to `SimpleCache` as you have suggested, so we can remove this Guava cache dependency.

RaulGracia commented:
Sure @shshashwat, system tests are green `test-pravega-continuous/4448`.

shshashwat reviewed (approved):
LGTM

RaulGracia commented:
@pbelgundi @andreipaduroiu can we wrap up this one?

pbelgundi commented on common/src/main/java/io/pravega/common/tracing/RequestTracker.java:
```diff
@@ -38,13 +39,15 @@
     private final Object lock = new Object();
     @GuardedBy("lock")
-    private final Cache<String, List<Long>> ongoingRequests;
+    private final SimpleCache<String, List<Long>> ongoingRequests;
```
What is the advantage of replacing google guava Cache with Simple Cache?
Google cache enables using Soft/Weak references for cache entries which enables fast reclamation of heap memory.

RaulGracia replied:
My understanding is that @andreipaduroiu suggested this to unify the usage of caches across Pravega, which makes sense. Anyway, irrespective of the implementation of Guava cache, without this PR we could get OOM anyway, and I think that with this PR we will be ok in terms of heap irrespective of the cache implementation used.

pbelgundi replied:
The objective of this PR is to reduce the amount of memory consumed by Request Tracker and I don;t see how this change would do that. I understand using the same library across components is a good idea but the changes in this PR need to focus on reducing the memory occupied by RequestTracker.

RaulGracia replied:
Ok, we can do the switch to `SimpleCache` separately.

pbelgundi commented:
PDP-31 states the intent of Request Tracker as:
"
to attach global tags to client requests that propagate through other Pravega components
"

However, in the code, I see that Request Ids generated by Request Tracker are almost never created on the Client, but are created inside ControllerServiceImpl which is a Controller (server side) class.
Assuming that the purpose of this class is to track all replays of the same request using the same request Id, I don;t see why we store a list of requestIds against a request descriptor:
```
[ongoingRequests](https://github.com/pravega/pravega/blob/a5f5d670b7f2340b8388a9a2b216aa7e5b4d7151/common/src/main/java/io/pravega/common/tracing/RequestTracker.java#L47)
```

RaulGracia commented:
@pbelgundi I addressed your comment restoring again the Guava cache and answered your comment. Let me know if there is anything else to address in this PR.

pbelgundi commented:
> @pbelgundi I addressed your comment restoring again the Guava cache and answered your comment. Let me know if there is anything else to address in this PR.

My comment wasn't so much about the use of Guava or Simple Cache, but more about what does it take to reduce the memory footprint of Request Tracker objects as in case of connectivity issues with Segment Store since these seem to be piling up in memory causing the Controller to go OOM.
We seem to be trying to limit the length of concurrent request id queue using - MAX_PARALLEL_REQUESTS (hard coded to 10) however this limit is already provided to Guava Cache by the "maximumSize" parameter and set to `1000000`.
Instead of adding entries to the cache and removing them, it may be better to just change the value of this parameter to a more reasonable setting, since I don't see why we would ever want to store as many request Ids in memory.

RaulGracia commented:
@pbelgundi I have addressed your comment and replied to your question. If you have no further observations, please merge this PR.

RaulGracia commented:
@pbelgundi any further comments on this PR?

pbelgundi commented:
> @pbelgundi the memory consumption in the `RequestTracker` has 2 dimensions:
> 
> 1. Number of **entries in the cache**, limited by `MAX_CACHE_SIZE=1000000`. This has been always limited.
> 2. Number of **request ids per entry** (i.e. request descriptor). This was not limited before and it is what this PR is addressing by adding `MAX_PARALLEL_REQUESTS=10`. Without the current change, it could happen that a Controller or a Client is retrying always the same action due to some issue (e.g., "create Stream X"). In a extreme case, this would result on having only 1 entry in the cache (therefore, the 1 million limit is respected and does not help), but the list of request ids in this cache entry would keep growing without bounds by adding new request ids. Now this is now fixed.
> 
> > Instead of adding entries to the cache and removing them, it may be better to just change the value of this parameter to a more reasonable setting, since I don't see why we would ever want to store as many request Ids in memory.
> 
> Well, I can set the limit of the `RequestTracker` to something lower (like 100K). But this is a different issue, because this would only impact on dimension 1, and what this PR fixes is dimension 2. Anyway, I will reduce the max amount entries to keep in cache, just in case it helps in any way to reduce memory (in an scenario of retries, it will make no difference, though).

Ok. I see. 
If we always use and log only the first request Id, what is the value in storing other request Ids? Wouldn't it help to store only 1 request Id per Request Descriptor?

RaulGracia commented:
@pbelgundi I have already answered that above:

> As I mentioned, in RequestTracker we have an idempotent Stream operation (create-stream-xxx) related to a client-generated request id yyy. In general, the cache will only contain a pair of these per operation. However, there could be cases in which multiple clients are attempting to do the same action and we want to track all of them (if possible). For instance, if we have 2 clients trying to create the same stream, the RequestTracker entry will look like: [create-stream-xxx: [yyy, zzz]. For consistency, we always pick the first request id associated to a request descriptor, as the Controller guarantees for idempotent operations guard us to be processing 2 operations for the same Stream. Therefore, we just care about the request id that is actually doing the processing and we keep using the same "primary" request id throughput its lifecycle (otherwise, we would be mixing multiple request ids and it would be pointless). The other outstanding request ids from other clients are also logged when requests are untracked (it may not be super useful, but it gives a bit more of information). Also, the point of having track of these parallel requests is to know when to "delete" the entry from the RequestTracker (if you only keep one, then you don't know when is the last request processed to remove that safely from the cache, unless you add a counter).

sachin-j-joshi commented on common/src/main/java/io/pravega/common/tracing/RequestTracker.java:
```diff
@@ -38,8 +38,10 @@
+    static final int MAX_PARALLEL_REQUESTS = 10;
     private static final String INTER_FIELD_DELIMITER = "-";
-    private static final int MAX_CACHE_SIZE = 1000000;
+    private static final int MAX_CACHE_SIZE = 100000;
```
Does it make sense to make it a configurable property ?

RaulGracia replied:
This has not been a problem in severa years this has been set to that value, and even though I just reduced it by 10x lower based on the reviews. We could create a parameter for that in a separate PR, but I doubt that this is going to be used by anyone nor solve the problem we are discussing in this PR.

sachin-j-joshi commented on common/src/main/java/io/pravega/common/tracing/RequestTracker.java:
```diff
@@ -167,10 +171,14 @@ public void trackRequest(String requestDescriptor, long requestId) {
             List<Long> requestIds = ongoingRequests.getIfPresent(requestDescriptor);
             if (requestIds == null) {
                 requestIds = Collections.synchronizedList(new ArrayList<>());
```
As we recently discovered `Collections.synchronizedList(new ArrayList<>());` is not thread safe when iterators are used. 
Please double check if we are using iterators on this collection anywhere. 
either add synchronized while using or change this to Vector something more thread safe.

RaulGracia replied:
There is no iteration over the list of `requestIds` associated to a request.

pbelgundi commented:
> @pbelgundi I have already answered that above:
> 
> > As I mentioned, in RequestTracker we have an idempotent Stream operation (create-stream-xxx) related to a client-generated request id yyy. In general, the cache will only contain a pair of these per operation. However, there could be cases in which multiple clients are attempting to do the same action and we want to track all of them (if possible). For instance, if we have 2 clients trying to create the same stream, the RequestTracker entry will look like: [create-stream-xxx: [yyy, zzz]. For consistency, we always pick the first request id associated to a request descriptor, as the Controller guarantees for idempotent operations guard us to be processing 2 operations for the same Stream. Therefore, we just care about the request id that is actually doing the processing and we keep using the same "primary" request id throughput its lifecycle (otherwise, we would be mixing multiple request ids and it would be pointless). The other outstanding request ids from other clients are also logged when requests are untracked (it may not be super useful, but it gives a bit more of information). Also, the point of having track of these parallel requests is to know when to "delete" the entry from the RequestTracker (if you only keep one, then you don't know when is the last request processed to remove that safely from the cache, unless you add a counter).

It is still not clear how "RequestTracker entry will look like: [create-stream-xxx: [yyy, zzz]" maps to different values for different clients. What is more than 10 clients place the same request on Controller at a the same time. With MAX_PARALLEL set to 10 will we be able to track the flows separately for each client

RaulGracia commented:
> It is still not clear how a RequestTracker entry like: [create-stream-xxx: [yyy, zzz] maps to different values for different clients. Which part is the clientId here? 

Client ids are each of the elements in the list `[yyy, zzz]`. The RequestTracker just tracks the action being done "create stream X" with the request ids from the clients doing so "`yyy, zzz`". So you can see what is the client doing the operation in the server, as well as the workflow of the operation.

> What if more than 10 clients place the same request on Controller at a the same time. With MAX_PARALLEL set to 10 would we be able to track the flows separately for each client

We have to put a limit to the max number of parallel request ids to limit the amount of memory the list of request ids will consume in case of continuous retries if the cluster is in bad state. If the cluster is in that bad state, I don't think the request ids are really relevant, because there is a problem in the server side or the environment that may cause the cluster to be in that state. In normal conditions, having more than 10 parallel requests means that there would be more than 10 clients doing the same operation on the same Stream (create, delete, etc), which normally take tens of milliseconds to complete. In the case this happens, for the additional parallel requests, we would probably lose track of them, as we are not allowing to keep their client request ids, and when untracking them from the cache it would return `RequestTag.NON_EXISTENT_ID`.
